### PR TITLE
Swatches from Color Inspector

### PR DESCRIPTION
**Pull request by milg15**
I am still figuring out the best way to change the SwatchId of the fill.

### CODE REVIEW

File: packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx [hunk 1]
```diff
@@ -150,12 +152,15 @@ export default memo(function ColorInputFieldWithPicker({
...
-  const [state] = useApplicationState();
-
+  const [state, dispatch] = useApplicationState();
   const [swatchLayout, setSwatchLayout] = useState<SwatchLayout>('grid');
 
   const sharedSwatches = Selectors.getSharedSwatches(state);
 
+  const isSwatch =
+    value.swatchID &&
+    sharedSwatches.some((e) => e.do_objectID === value.swatchID);
```
Comment (dabbott) [inline packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx:162]: I think theoretically we shouldn't need to check `sharedSwatches` at all, cause all `swatchID` should be valid if they exist? Did you find this was needed?

File: packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx [hunk 2]
```diff
@@ -168,6 +173,26 @@ export default memo(function ColorInputFieldWithPicker({
...
+            onClick={useCallback(() => {
+              const id = uuid();
+              onChange({
+                ...value,
+                swatchID: isSwatch ? undefined : id,
+              });
+              if (isSwatch) return;
+
+              const swatchName = prompt('New Swatch Name');
+              if (!swatchName) return;
```
Comment (dabbott) [inline packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx:189]: We should call `onChange` after we know we have a swatchName, so we don't end up in an inconsistent state

File: packages/noya-designsystem/src/components/Button.tsx [hunk 3]
```diff
@@ -11,7 +11,8 @@ type ButtonVariant = 'normal' | 'thin';
 const ButtonElement = styled.button<{
   active: boolean;
   variant: ButtonVariant;
-}>(({ theme, active, disabled, variant }) => ({
+  full: boolean;
```
Comment (dabbott) [inline packages/noya-designsystem/src/components/Button.tsx:14]: We might need to add this, but if we do, I think we should try to use flexbox.

In this case though, I think we can get the right layout by making `PaddedSection` a flex container:

Screen Shot 2021-06-03 at 11 08 23 AM

**Review by dabbott — CHANGES_REQUESTED** [review]
I tried it and it looks great 💯

File: packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx (hunk 1 above)
Comment (milg15) [reply packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx:162]: Maybe if we delete a swatch but they color still has the swatchId? 🤔

Comment (dabbott) [reply packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx:162]: In that case, we should probably update our swatch deletion code to traverse layers/styles and delete the swatchIDs. Similar to how we do `visitLayerColors` and `visitStyleColors` with "setSwatchColor"

File: packages/noya-designsystem/src/components/Button.tsx [hunk 4]
```diff
@@ -62,6 +62,7 @@ interface ButtonRootProps {
   disabled?: boolean;
   variant?: ButtonVariant;
   tooltip?: ReactNode;
+  full?: boolean;
```
Comment (dabbott) [inline packages/noya-designsystem/src/components/Button.tsx:65]: Remove

File: packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx (hunk 1 above)
Comment (dabbott) [reply packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx:164]: Do you want to update the deletion code in this PR? Or should we do it another time?

**Review by dabbott — APPROVED** [review]

**Comment by milg15** [comment]
How update the deletion ?

**Comment by dabbott** [comment]
Currently when we `setSwatchColor`, we update every instance of that color with the same swatchID.

Screen Shot 2021-06-03 at 12 25 34 PM

We need to do the same when removing a swatch. For any color with the same swatchID, we need to set the swatchID of that color to undefined.

Screen Shot 2021-06-03 at 12 26 24 PM

**Comment by milg15** [comment]
Oh! I can add it to this PR actually

File: packages/noya-state/src/reducers/themeReducer.ts [hunk 5]
```diff
@@ -505,6 +529,19 @@ export function themeReducer(
...
         draft.sketch.document.sharedSwatches = sharedSwatches;
+
+        layerIndexPathsWithSwatchId.forEach((layerPath) =>
+          accessPageLayers(
+            draft,
+            layerPath.pageIndex,
+            layerPath.indexPaths,
+          ).forEach((layer) => {
+            layer.style?.fills?.forEach((f) => {
+              if (f.color.swatchID && ids.includes(f.color.swatchID))
```
Comment (dabbott) [inline packages/noya-state/src/reducers/themeReducer.ts:540]: Very minor but we already checked this condition above, so I think it should be safe to just `delete` the swatch without the condition. Anything in `layerIndexPathsWithSwatchId` should have the correct swatchID

File: packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx (hunk 1 above)
Comment (dabbott) [reply packages/app/src/components/inspector/ColorInputFieldWithPicker.tsx:164]: Now that we've updated the swatch deletion, can we delete this `.some` check?

File: packages/noya-state/src/reducers/themeReducer.ts (hunk 5 above)
Comment (milg15) [reply packages/noya-state/src/reducers/themeReducer.ts:540]: The layer can have multiple fills

Comment (dabbott) [reply packages/noya-state/src/reducers/themeReducer.ts:540]: Oh yeah, good point. Ok, keep it the way it is

**Pull request merged**